File: src/steam.py
import os
import winreg
from typing import List
from pathlib import Path

def find_steam_folders() -> List[str]:
    """
    Searches the system for Steam installation and library folders.
    Returns a list of paths where Steam folders are found.
    """
    steam_paths = set()
    
    # Common Steam installation paths to check
    default_paths = [
        "C:\\Program Files (x86)\\Steam",
        "C:\\Program Files\\Steam",
    ]
    
    # Try to get Steam path from Windows Registry
    try:
        hkey = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, "SOFTWARE\\WOW6432Node\\Valve\\Steam")
        reg_path = winreg.QueryValueEx(hkey, "InstallPath")[0]
        if os.path.exists(reg_path):
            steam_paths.add(reg_path)
    except WindowsError:
        pass

    # Check default installation paths
    for path in default_paths:
        if os.path.exists(path):
            steam_paths.add(path)
    
    # Look for additional library folders in libraryfolders.vdf
    for steam_path in steam_paths.copy():
        library_file = os.path.join(steam_path, "steamapps", "libraryfolders.vdf")
        if os.path.exists(library_file):
            with open(library_file, 'r', encoding='utf-8') as f:
                content = f.read()
                # Simple parsing of VDF format to find paths
                for line in content.splitlines():
                    if '"path"' in line.lower():
                        # Extract path between quotes
                        path = line.split('"')[3].replace("\\\\", "\\")
                        if os.path.exists(path):
                            steam_paths.add(path)
    
    return list(steam_paths)

if __name__ == "__main__":
    # Test the function
    steam_folders = find_steam_folders()
    print("Found Steam folders:")
    for folder in steam_folders:
        print(f"- {folder}")

File: src/steam.cpp
#include <windows.h>
#include <string>
#include <vector>
#include <set>
#include <fstream>
#include <filesystem>
#include <iostream>

std::vector<std::string> find_steam_folders() {
    std::set<std::string> steam_paths;
    
    // Search drives for Steam installations
    for (char drive = 'A'; drive <= 'Z'; drive++) {
        std::string base_path = std::string(1, drive) + ":\\";
        // Remove the DRIVE_FIXED check to search all drive types
        if (GetDriveTypeA(base_path.c_str()) == DRIVE_NO_ROOT_DIR) continue;
        
        try {
            // Check root Steam folder
            std::string steam_path = base_path + "Steam";
            if (std::filesystem::exists(steam_path + "\\steam.exe")) {
                steam_paths.insert(steam_path);
                continue;
            }

            // Check one level deep
            for (const auto& entry : std::filesystem::directory_iterator(base_path)) {
                if (!entry.is_directory()) continue;
                steam_path = entry.path().string() + "\\Steam";
                if (std::filesystem::exists(steam_path + "\\steam.exe")) {
                    steam_paths.insert(steam_path);
                }
            }
        } catch (const std::filesystem::filesystem_error&) {
            continue;
        }
    }
    // Get additional paths from libraryfolders.vdf
    for (const auto& path : steam_paths) {
        std::string vdf_path = path + "\\steamapps\\libraryfolders.vdf";
        if (!std::filesystem::exists(vdf_path)) continue;
        
        std::ifstream file(vdf_path);
        std::string line;
        while (std::getline(file, line)) {
            if (line.find("\"path\"") == std::string::npos) continue;
            
            size_t first = line.find('\"', line.find('\"') + 1) + 1;
            size_t last = line.find('\"', first);
            std::string lib_path = line.substr(first, last - first);
            
            // Normalize path
            size_t pos = 0;
            while ((pos = lib_path.find("\\\\", pos)) != std::string::npos) {
                lib_path.replace(pos, 2, "\\");
                pos++;
            }
            
            if (std::filesystem::exists(lib_path)) {
                steam_paths.insert(lib_path);
            }
        }
    }
    
    return std::vector<std::string>(steam_paths.begin(), steam_paths.end());
}

int main() {
    // Test the function
    std::vector<std::string> steam_folders = find_steam_folders();
    std::cout << "Found Steam folders:\n";
    for (const auto& folder : steam_folders) {
        std::cout << "- " << folder << "\n";
    }
    std::cout << "\nPress Enter to exit...";
    std::cin.get();
    return 0;
}